hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/SaslMechanismFactory.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.security;
+
+import org.apache.hadoop.HadoopIllegalArgumentException;
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.hadoop.fs.CommonConfigurationKeysPublic.HADOOP_SECURITY_SASL_MECHANISM_DEFAULT;
+import static org.apache.hadoop.fs.CommonConfigurationKeysPublic.HADOOP_SECURITY_SASL_MECHANISM_KEY;
+
+/**
+ * SASL related constants.
+ */
+@InterfaceAudience.LimitedPrivate({"HDFS", "MapReduce"})
+@InterfaceStability.Evolving
+public final class SaslMechanismFactory {
+  static final Logger LOG = LoggerFactory.getLogger(SaslMechanismFactory.class);
+
+  private static final String SASL_MECHANISM_ENV = "HADOOP_SASL_MECHANISM";
+  private static final String SASL_MECHANISM;
+
+  static {
+    // env
+    final String envValue = System.getenv(SASL_MECHANISM_ENV);
+    LOG.debug("{} = {} (env)", SASL_MECHANISM_ENV, envValue);
+
+    // conf
+    final Configuration conf = new Configuration();
```

Comment:
I am a little concerned about this one.
If an application loads configuration from a custom path, SaslMechanismFactory initialized this way will load configurations from default path, and results in unexpected behavior to the user.

And that is a potential security issue. I had some grief over similar problems before, see: [HADOOP-13638](https://issues.apache.org/jira/browse/HADOOP-13638)

Reply:
I seems to recall similar problem so that I did not use conf previously.  Cc @stoty 

@jojochuang , If we use `new Configuration(false)`, the value is not set when the application loads configuration from a custom path.  That application must set env variable.  Is it okay?

Reply:
I don't follow the issue here.
Is it about adding config files to the classpath dynamically ?
Is SaslMechanismFactory reading the Configuration too early ?

It should be possble (with more changes) to load the mechanism lazily, when SASL is used first.

IMO having to configure this from an environment variable would be an operational nightmare.

Reply:
> It should be possble (with more changes) to load the mechanism lazily, when SASL is used first.

The is already loaded mechanism lazily.  The class won't be loaded unless it is used.

In our case, it should be fine since conf will be loaded during a SASL server/client is being created.  The conf path must be set at that time.  (`UserGroupInformation` has a different story since it may have to logged in before starting a server/client.)

> IMO having to configure this from an environment variable would be an operational nightmare.

Hadoop security tends to use environment variable since conf may not available.

Reply:
The rest of PR looks good to me. This part I'm not sure. So what about the normal case, like YARN Resource Manager, do they also get the value from env variable? Then what's the use of the configuration object?

Reply:
@jojochuang , @stoty , unfortunately, UserGroupInformation is statically referring to `SaslMechanismFactory` (via `SaslRpcServer.AuthMethod`) as @steveloughran discovered.  We may need to remove the conf support from `SaslMechanismFactory`.

Reply:
We are able to use reflection to implement lazy loading of `SaslMechanismFactory`; see https://github.com/apache/hadoop/pull/7173